lib/galaxy/webapps/galaxy/services/wes.py

```diff
@@ -0,0 +1,1180 @@
+import json
+import logging
+from dataclasses import dataclass
+from typing import (
+    Any,
+    List,
+    Optional,
+)
+from urllib.parse import (
+    parse_qs,
+    urlparse,
+)
+
+from fastapi import UploadFile
+from sqlalchemy import (
+    literal,
+    select,
+    tuple_,
+    union_all,
+)
+from sqlalchemy.orm import joinedload
+
+from galaxy import exceptions
+from galaxy.config import GalaxyAppConfiguration
+from galaxy.files.uris import stream_url_to_str
+from galaxy.managers.context import ProvidesUserContext
+from galaxy.model.keyset_token_pagination import (
+    KeysetPagination,
+    SingleKeysetToken,
+)
+from galaxy.managers.workflows import (
+    RawWorkflowDescription,
+    WorkflowContentsManager,
+    WorkflowCreateOptions,
+)
+from galaxy.model import (
+    History,
+    ImplicitCollectionJobs,
+    ImplicitCollectionJobsJobAssociation,
+    Job,
+    WorkflowInvocation,
+    WorkflowInvocationStep,
+)
+from galaxy.schema.wes import (
+    DefaultWorkflowEngineParameter,
+    RunId,
+    RunListResponse,
+    RunLog,
+    RunRequest,
+    RunStatus,
+    RunSummary,
+    ServiceInfo,
+    State,
+    TaskListResponse,
+    TaskLog,
+    WorkflowEngineVersion,
+    WorkflowTypeVersion,
+)
+from galaxy.schema.workflows import InvokeWorkflowPayload
+from galaxy.security.idencoding import IdEncodingHelper
+from galaxy.webapps.galaxy.services.base import ServiceBase
+from galaxy.webapps.galaxy.services.ga4gh import build_service_info
+from galaxy.webapps.galaxy.services.workflows import WorkflowsService
+from galaxy.work.context import SessionRequestContext
+
+log = logging.getLogger(__name__)
+
+# Map Galaxy workflow invocation states to WES states
+GALAXY_TO_WES_STATE = {
+    "new": State.QUEUED,
+    "ready": State.INITIALIZING,
+    "scheduled": State.RUNNING,
+    "failed": State.EXECUTOR_ERROR,
+    "cancelled": State.CANCELED,
+    "cancelling": State.CANCELING,
+}
+
+WES_TO_GALAXY_STATE = {v: k for k, v in GALAXY_TO_WES_STATE.items()}
+
+
+@dataclass
+class TaskKeysetToken:
+    """Composite keyset token for task pagination (step_order, job_index).
+
+    Used to identify position in task list for cursor-based pagination.
+    """
+
+    step_order: int
+    job_index: int
+
+    def to_values(self) -> list:
+        """Convert token to normalized list of values for encoding."""
+        return [self.step_order, self.job_index]
+
+    @classmethod
+    def from_values(cls, values: list) -> "TaskKeysetToken":
+        """Reconstruct token from decoded values."""
+        if len(values) < 2:
+            raise ValueError("TaskKeysetToken requires at least 2 values")
+        return cls(step_order=values[0], job_index=values[1])
+
+
+def _parse_gxworkflow_uri(workflow_url: str) -> tuple[str, bool]:
+    """Parse a gxworkflow:// URI to extract workflow ID and instance flag.
+
+    Format: gxworkflow://<encoded_workflow_id>[?instance=<false|true>]
+    - encoded_workflow_id: Base62-encoded Galaxy workflow ID
```

Comment:
```suggestion
    - encoded_workflow_id: Base64-encoded Galaxy workflow ID
```
I guess in general I'm not sure I understand the need for base64 encoding if entities also have a meaning in the existing Galaxy API.

Should we mandate `instance=true` ? It is probably only known by dev's that the stored workflow may have a different active workflow association